### PR TITLE
[DEMO] QML for Phase Detection

### PR DESCRIPTION
**Pull request by SaverioMonaco**
**Title:**
Supervised and Unsupervised Quantum Machine Learning models for the phase detection of the ANNNI spin model

**Summary:**
Showcase of two foundational models in QML for Phase Detection for Spin Systems. One follows a supervised approach, the other follows an unsupervised approach.

**Relevant references:**

**Possible Drawbacks:**
QCNN models is considered now outdated.

**Related GitHub Issues:**

----
If you are writing a demonstration, please answer these questions to facilitate the marketing process.

* GOALS — Why are we working on this now?

  *Eg. Promote a new PL feature or show a PL implementation of a recent paper.*
Show a working example of the founding models in QML for Phase Detection. Show how MPS and QML can be used together.

* AUDIENCE — Who is this for?

  *Eg. Chemistry researchers, PL educators, beginners in quantum computing.*
Beginners in Quantum Computing, Chemistry researchers

* KEYWORDS — What words should be included in the marketing post?
ANNNI, QCNN, QAD, Phase Detection

* Which of the following types of documentation is most similar to your file? 
(more details [here](https://www.notion.so/xanaduai/Different-kinds-of-documentation-69200645fe59442991c71f9e7d8a77f8))
    
- [ x] Tutorial
- [ ] Demo
- [ ] How-to

[sc-80710]

### CODE REVIEW

**Comment by SaverioMonaco** [comment]
- The image `annni_sketch.png` is a rough sketch for the thumbnail. It needs to be updated by replacing $J2$ with $\kappa$ and $J$ with $J_1$ to ensure consistency with the Hamiltonian described in Eq. (1).  

- The other two images, `annni_pd.png` and `annni_pd_analytical.png`, were originally generated using `matplotlib.pyplot`. I imported them as images to reduce code volume. I will add the code to reproduce these images shortly.

**Comment by SaverioMonaco** [comment]
- `annni_pd.png` 

```python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap, BoundaryNorm

# Kosterlitz-Thouless transition line
def kt_transition(k):
    return 1.05 * np.sqrt((k - 0.5) * (k - 0.1))

# Ising transition line
def ising_transition(k):
    return np.where(k == 0, 1, (1 - k) * (1 - np.sqrt((1 - 3 * k + 4 * k**2) / (1 - k))) / np.maximum(k, 1e-9))

# Floating Phase transition line
def bkt_transition(k):
    return 1.05 * (k - 0.5)
    
# Get the phase from the DMRG transition lines
def get_phase(k, h):
    if k < .5 and h < ising_transition(k):
        return 0
    elif k > .5 and h < kt_transition(k):
        return 1
    return 2

# Generate the phase diagram data from DMRG transition lines
ks100, hs100 = np.linspace(0, 1, 100), np.linspace(0, 2, 100)
K100, H100 = np.meshgrid(ks100, hs100)
img_dmrg = np.vectorize(get_phase)(K100, H100)

colors = ['#80bfff', '#fff2a8',  '#80f090', '#da8080',]
phase_labels = ["Ferromagnetic", "Antiphase", "Paramagnetic", "Trash Class",]
cmap = ListedColormap(colors)

bounds = [-0.5, 0.5, 1.5, 2.5, 3.5]
norm = BoundaryNorm(bounds, cmap.N)

# Plot the phase diagram
plt.figure(figsize=(4, 4))
plt.imshow(img_dmrg, cmap=cmap, aspect="auto", origin="lower", extent=[0, 1, 0, 2], norm=norm)

# Plot the transition lines.
k_vals1 = np.linspace(0.0, 0.5, 50)
k_vals2 = np.linspace(0.5, 1.0, 50)
plt.plot(k_vals1, ising_transition(k_vals1), 'k', lw=2)
plt.plot(k_vals2, kt_transition(k_vals2), 'k', lw=2)
plt.plot(k_vals2, bkt_transition(k_vals2), 'k', ls = '--', lw=2)

# Create legend entries
for color, phase in zip(colors, phase_labels[:-1]):
    plt.scatter([], [], color=color, label=phase, edgecolors='black')
plt.plot([], [], 'k', label='Transition Lines')
plt.text(0.15, 0.5, s=r'$h_I$', fontsize=12)
plt.text(0.70, 0.5, s=r'$h_C$', fontsize=12)
plt.text(0.8, 0.2, s=r'$h_{BKT}$', fontsize=12)

plt.legend(), plt.xlabel("k"), plt.ylabel("h"), plt.title("Phase diagram"), plt.show();

```

- `annni_pd_analytical.png`

```python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap, BoundaryNorm

# Kosterlitz-Thouless transition line
def kt_transition(k):
    return 1.05 * np.sqrt((k - 0.5) * (k - 0.1))

# Ising transition line
def ising_transition(k):
    return np.where(k == 0, 1, (1 - k) * (1 - np.sqrt((1 - 3 * k + 4 * k**2) / (1 - k))) / np.maximum(k, 1e-9))

# Floating Phase transition line
def bkt_transition(k):
    return 1.05 * (k - 0.5)
    
# Get the phase from the DMRG transition lines
def get_phase(k, h):
    if k < .5 and h < ising_transition(k):
        return 0
    elif k > .5 and h < kt_transition(k):
        return 1
    return 2

# Generate the phase diagram data from DMRG transition lines
ks100, hs100 = np.linspace(0, 1, 100), np.linspace(0, 2, 100)
K100, H100 = np.meshgrid(ks100, hs100)
img_dmrg = np.vectorize(get_phase)(K100, H100)

colors = ['#80bfff', '#fff2a8',  '#80f090', '#da8080',]
phase_labels = ["Ferromagnetic", "Antiphase", "Paramagnetic", "Trash Class",]
cmap = ListedColormap(colors)

bounds = [-0.5, 0.5, 1.5, 2.5, 3.5]
norm = BoundaryNorm(bounds, cmap.N)

# Plot the phase diagram
plt.figure(figsize=(4, 4))
plt.imshow(img_dmrg, cmap=cmap, aspect="auto", origin="lower", extent=[0, 1, 0, 2], norm=norm)

# Plot the transition lines.
k_vals1 = np.linspace(0.0, 0.5, 50)
k_vals2 = np.linspace(0.5, 1.0, 50)
plt.plot(k_vals1, ising_transition(k_vals1), 'k', lw=2)
plt.plot(k_vals2, kt_transition(k_vals2), 'k', lw=2)
plt.plot(k_vals2, bkt_transition(k_vals2), 'k', ls = '--', lw=2)

# Create legend entries
for color, phase in zip(colors, phase_labels[:-1]):
    plt.scatter([], [], color=color, label=phase, edgecolors='black')
plt.plot([], [], 'k', label='Transition Lines')

# Highlight analytical points (k = 0 and h = 0)
plt.scatter(np.zeros_like(hs100), hs100, color='red', s=20, label="Analytical Points")  # y-axis points
plt.scatter(ks100, np.zeros_like(ks100), color='red', s=20)  # x-axis point

plt.text(0.15, 0.5, s=r'$h_I$', fontsize=12)
plt.text(0.70, 0.5, s=r'$h_C$', fontsize=12)
plt.text(0.8, 0.2, s=r'$h_{BKT}$', fontsize=12)

plt.legend(), plt.xlabel("k"), plt.ylabel("h"), plt.title("Phase diagram"), plt.show();

```

**Review by daniela-angulo — CHANGES_REQUESTED** [review]
This was a joy to read. I added a few comments/suggestions.

File: demonstrations/tutorial_annni.py [hunk 1]
```diff
@@ -0,0 +1,705 @@
...
+    :target: javascript:void(0)
+
+    Figure 1. Add a caption here
+
+The ANNNI model
+--------------------------------------------------------------------------------
+
+The ANNNI model describes a spin system with three types of competing interactions. Its Hamiltonian is given by:
+
+.. math::  H =   -J \sum_{i=1}^{N} \sigma_x^i\sigma_x^{i+1} - \kappa \sigma_x^{i}\sigma_x^{i+2} + h \sigma_z^i \tag{1}
```
[the remark this answers is not in the document]
Comment (SaverioMonaco) [reply demonstrations/tutorial_annni.py]: The signs in the ANNNI Hamiltonian have always been a source of confusion. The important thing is getting the relative sign between the nearest-neighbor and next-nearest-neighbor interactions right, otherwise, there is no competition between them, which leads to no actual phase transition (I know this firsthand, it cost me a week of debugging at the time).

The [original (1988) paper](https://www.sciencedirect.com/science/article/pii/0370157388901408) defines the hamiltonian as (Eq. 2.1): 
$$H = - J_1\Big(\sum_i \sigma_x^i \sigma_x^{i+1} - \big( \underbrace{-\frac{J_2 }{J_1}}_{\kappa} \big) \sigma_x^i \sigma_x^{i+1} ... \Big) =  - J_1\Big(\sum_i \sigma_x^i \sigma_x^{i+1} - \kappa \sigma_x^i \sigma_x^{i+1} ... \Big)$$
For $J_1 >  0$ and $J_2 \leq 0$ hence $\kappa \geq 0$

In our paper we defined it as (Eq. 1) 
$$H =   J_1 \sum_i \sigma_x^i\sigma_x^{i+1} - \kappa \sigma_x^{i}\sigma_x^{i+2} + h \sigma_z^i$$

I assume in our paper we forgot to add the minus sign before the nearest-neighbor term. Thank you for pointing that out! 

I would personally leave as it is the equation in the notebook, as it is the same formulation of the original 1988 paper.

File: demonstrations/tutorial_annni.py [hunk 2]
```diff
@@ -0,0 +1,705 @@
...
+#
+# Unsupervised learning of phases: Quantum Anomaly Detection
+# ----------------------------------------------------------
+# 
+# Quantum Anomaly Detection (QAD) was first introduced in [#Kottmann]_ and serves as the quantum counterpart of an Auto-Encoder, with only the encoding (forward) process being trained (due to the inversion property of quantum unitaries).
+# 
+# In this approach, a single state :math:`|\psi(\kappa, h)\rangle` from the ANNNI model is optimised to find the best circuit parameters such that:
+# 
+# 
+# .. math::  V(\theta)|\psi(\kappa, h)\rangle = |\phi\rangle^{N-K} \otimes |0\rangle^{\otimes K}\tag{6}
```
[the remark this answers is not in the document]
Comment (SaverioMonaco) [reply demonstrations/tutorial_annni.py:454]: I rephrased the whole paragraph in the commit, I hope it is now clearer!

**Review by SaverioMonaco — COMMENTED** [review]
Final touches: 

- Acknowledgement to DESY
- Training of the QCNN with the full training set instead of sampling each epoch for a smoother loss curve
- Adapted hyperparameters of the training function of the QCNN

File: demonstrations/tutorial_annni.py [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+#    
+#    .. math::  \mathcal{C} = \frac{1}{2}\sum_{j\in q_T} (1-\left<Z_j\right>),\tag{7}
+#    
+#    where :math:`q_T` refers to the set of trash qubits, which make up :math:`N/2` of the total.
+#    By doing so, all the information of the input quantum state is compressed in the remaining non-measured qubits.
+#   
+# In this case, the selected quantum state corresponds to the trivial case with :math:`\kappa = 0` and :math:`h = 0`.
+
+def train_anomaly(num_epochs, lr, seed):
+    """Training function of the QCNN architecture"""
```
Comment (SaverioMonaco) [inline demonstrations/tutorial_annni.py:527]:
```suggestion
    """Training function of the QAD architecture"""
```

**Review by SaverioMonaco — COMMENTED** [review]
Fix of a comment in the QAD training function

File: demonstrations/tutorial_annni.py (hunk 3 above)
Comment (CatalinaAlbornoz) [reply demonstrations/tutorial_annni.py:527]: Good catch!

**Review by daniela-angulo — APPROVED** [review]
